Bugfix/agilis 4190 limiting management to only selected interfaces

metelik opened the pull request:
Well adding user-scenario of limiting the network interfaces managed by nerves network and nerves network interface to an explicitly specified list.

Commit:
Adding configuring the list of interfaces to be managed

Commit:
Adding configuring the list of interfaces to be managed

Commit:
Adding configuration for the list of interfaces to be managed

Commit:
Adding the allmulti flag to the network interface

alanj853 reviewed (approved):
Perhaps change that string that we talked about to an atom, but other than that it looks good.

metelik committed:
Update netif.c

Testing solution to problems like:
'usr/include/linux/if.h:143:8: error: redefinition of ‘struct ifmap’
 struct ifmap {
        ^'

metelik committed:
Update netif.c

revoking the test change

adkron commented on lib/network_interface/worker.ex:
```diff
@@ -125,8 +126,37 @@ defmodule Nerves.NetworkInterface.Worker do
   end
 
+  #Returns intersection of lists a and b
+  defp intersect(a, b), do: a -- (a -- b)
```
Do we need to remove duplicates too? This function doesn't understand duplicated.

The idiomatic way is `MapSet.intersection(MapSet.new(a), MapSet.new(b)) |> MapSet.to_list()`. I don't know if it is faster or better on memory. I thought you might want it if you are dealing with possible duplicates. I do know that the MapSet version does a sort before hand so that it doesn't have to walk the entire list. I don't think the list version does that.

adkron commented on lib/network_interface/worker.ex:
```diff
@@ -125,8 +126,37 @@ defmodule Nerves.NetworkInterface.Worker do
+      get_managed_interfaces(available_interfaces)
+        |> intersect(available_interfaces)
+    Logger.debug "#{__MODULE__}: response: #{inspect response}"
+    IO.puts "#{__MODULE__}: response: #{inspect response}"
```
We need to remove the `puts` and rely on the logger.

adkron commented on lib/network_interface/worker.ex:
```diff
@@ -125,8 +126,37 @@ defmodule Nerves.NetworkInterface.Worker do
+    response =
+      get_managed_interfaces(available_interfaces)
+        |> intersect(available_interfaces)
+    Logger.debug "#{__MODULE__}: response: #{inspect response}"
```
What is the upper bound to the `response` list size? If it is of any significant size, we want to wrap the debug message in an `fn`. I'm not imagining this being that big, but I will defer to you since you've been working on this a lot. I'm not sure at which point the crossover of speed is. You can test it with `:timer.tc`.

adkron reviewed (approved):
The PR looks nice. Good work! Live long and prosper.

adrigonzo commented on config/config.exs:
```diff
@@ -0,0 +1,9 @@
+use Mix.Config
+
+#config :nerves_network_interface, :managed_interfaces,
```
howcome these are commented out?

adrigonzo commented on src/netif.c:
```diff
@@ -1584,6 +1587,7 @@ static int remove_all_gateways6(struct netif *nb, const char *ifname)
     /* There may be more than one gateway. Remove all of them. */
     for (;;) {
         int rc = ioctl(nb->inet6_fd, SIOCDELRT, &route);
+        debug("Removing GW returnt rc = %d\r\n", rc);
```
just typo

adrigonzo commented on lib/network_interface/worker.ex:
```diff
@@ -125,8 +126,37 @@ defmodule Nerves.NetworkInterface.Worker do
+    end
+  end
+
+  def handle_call(:all_interfaces, _from, state) do
```
for all your protesting, you managed to have nice small functions! ;) looks good!
